reactor-core/src/main/java/reactor/core/publisher/Flux.java

```diff
@@ -2387,7 +2387,12 @@ public final T blockFirst() {
 	public final T blockFirst(Duration timeout) {
 		BlockingFirstSubscriber<T> subscriber = new BlockingFirstSubscriber<>();
 		subscribe((Subscriber<T>) subscriber);
-		return subscriber.blockingGet(timeout.toMillis(), TimeUnit.MILLISECONDS);
+		if (Operators.nanoPrecision(timeout)) {
```

Comment:
maybe I am missing something, but why not always `blockingGet(timeout.toNanos(), TimeUnit.NANOSECONDS)`?

Reply:
https://github.com/reactor/reactor-core/issues/1734#issuecomment-499405676

@simonbasle mentioned a check for MAX_NANOS, so I stuck with it.

If you have `Duration.ofNanos(Long.MAX_VALUE).plussNanos(1) ` and you try to convert to a `long` with `toNanos()`, you will get an on overflow exception. The check allows for this by using millisecond precision instead in this case.

That being said, Duration.ofNanos(Long.MAX_VALUE) is equivalent to 292 years, as he pointed out, and Duration.ofMillis(Long.MAX_VALUE) is equivalent to 292 million years. So, we can probably do without the check for MAX_NANOS, unless there's a good reason to  preserve the same ceiling of 292 million years for time (instead of a measly 292 years).

Reply:
yeah, blocking for 292 years already sounds like a bad thing 😜
I think keeping a nano precision for blocking methods is sane, could be documented though. and also, these methods produce error messages that mention the duration, which doesn't look terribly good when using `Duration#toString` 😞

Reply:
Would this not apply to all of the duration-based operators? What cases are we accounting for when doing this in the operators?:

```
if (Operators.nanoPrecision(timeout)) {
    return subscriber.blockingGet(timeout.toNanos(), TimeUnit.NANOSECONDS);
}
else {
    return subscriber.blockingGet(timeout.toMillis(), TimeUnit.MILLISECONDS);
}
```

What if we only use nano-precision, and let the `ArithmeticException` occur for a provided `Duration` exceeding 292 years?

Otherwise, if you check for `MAX_NANOS`, a user can still technically provide `Duration.ofSeconds(Long.MAX_VALUE)`, and that will overflow in both milli and nanoseconds.

Maybe something like this?

```
try {
    this.timespan = duration.toNanos();
}
catch (ArithmeticException e) {
    this.timespan = Long.MAX_VALUE;
}
```

Reply:
yeah, the more I think about it, the less I can find a reason not to use NANOS precision in all cases. Even `delay*` operators don't make much sense past 292 years... This is unlikely to change, whereas _gaining_ precision in the JVM's clock and scheduling primitives sound more likely in the near(ish) future.

any opinion / insight, @reactor/discuss ?